build-tools/packages/build-cli/src/BasePackageCommand.ts

```diff
@@ -0,0 +1,164 @@
+		}),
+		...BaseCommand.flags,
+	};
+
+	protected abstract processPackage(directory: string): Promise<void>;
+
+	private async processPackages(directories: string[]): Promise<void> {
+		const scopeIn = scopesToPrefix(this.flags.scope);
+		const scopeOut = scopesToPrefix(this.flags.skipScope);
+
+		const packages = directories.filter((directory) => {
+			const json: PackageJson = readJSONSync(path.join(directory, "package.json"));
+			const isPrivate: boolean = json.private ?? false;
+			if (this.flags.private !== undefined && this.flags.private !== isPrivate) {
+				return false;
+			}
+			if (scopeIn !== undefined) {
+				let found = false;
+				for (const scope of scopeIn) {
+					found ||= json.name?.startsWith(scope) ?? false;
+				}
+				if (!found) return false;
+			}
+			if (scopeOut !== undefined) {
+				for (const scope of scopeOut) {
+					if (json.name?.startsWith(scope) ?? false) {
+						return false;
+					}
+				}
+			}
+			return true;
+		});
+
+		let started = 0;
+		let finished = 0;
+		let succeeded = 0;
+		// In verbose mode, we output a log line per package. In non-verbose mode, we want to display an activity
+		// spinner, so we only start the spinner if verbose is false.
+		const verbose = this.flags.verbose;
+
+		if (verbose) {
+			this.info(`Filtered ${listNames(directories)} packages to ${listNames(packages)}`);
+		}
+
+		function updateStatus(): void {
+			if (!verbose) {
+				ux.action.start(
+					"Processing Packages...",
+					`${finished}/${packages.length}: ${started - finished} pending. Errors: ${
+						finished - succeeded
+					}`,
+					{
+						stdout: true,
+					},
+				);
+			}
+		}
+		try {
+			await async.mapLimit(packages, 25, async (directory) => {
+				started += 1;
+				updateStatus();
+				try {
+					await this.processPackage(directory);
+					succeeded += 1;
+				} finally {
+					finished += 1;
+					updateStatus();
+				}
+			});
+		} finally {
+			// Stop the spinner if needed.
+			if (!verbose) {
+				ux.action.stop(`Done. ${packages.length} Packages. ${finished - succeeded} Errors`);
+			}
+		}
+	}
+
+	public async run(): Promise<void> {
+		const flags = this.flags;
+
+		const releaseGroup = flags.releaseGroup;
+		const independentPackages = flags.packages;
+		const dir = flags.dir;
+
+		if (dir !== undefined) {
+			return this.processPackages([dir]);
+		}
+
+		if (flags.releaseGroup === undefined && flags.packages === false) {
+			return this.processPackages(["."]);
+		}
+
+		const ctx = await this.getContext();
+		if (releaseGroup !== undefined) {
+			this.info(`Finding packages for release group: ${releaseGroup}`);
+			return this.processPackages(
+				ctx.packagesInReleaseGroup(releaseGroup).map((p) => p.directory),
+			);
+		}
+		assert(independentPackages);
+		this.info(`Finding independent packages`);
+		return this.processPackages(ctx.independentPackages.map((p) => p.directory));
+	}
+}
+
+function scopesToPrefix(scopes: undefined | string[]): string[] | undefined {
```

Comment:
```suggestion
function scopesToPrefix(scopes: string[] | undefined): string[] | undefined {
```

total nit, feel free to ignore